File: todoApi/src/TodoModule/Repositories/Todo/Repository/TodoRepository.cs
using todoApi.src.Repositories.Todo.Context;
using todoApi.src.Repositories.Todo.DTOs;

namespace todoApi.src.Repositories.Todo.Repository
{
    public class TodoRepository : ITodoRepository
    {
        private readonly TodoContext _todoContext;
        public TodoRepository(TodoContext todoContext)
        {
            _todoContext = todoContext;
        }
        public Models.Todo Create(Models.Todo todo)
        {
            var newTodo = new Entity.TodoEntity(todo.Description, todo.CreatedAt, todo.Status, todo.CompletionDate);
            try
            {
                _todoContext.Add(newTodo);
                _todoContext.SaveChanges();
            }
            catch (Exception)
            {
                throw;
            }
            todo.Id = newTodo.Id;
            return todo;
        }

        public bool Delete(int id)
        {
            var todo = _todoContext.Todos.Where((id) => id.Equals(id)).FirstOrDefault();
            if (todo == null)
                return false;
            try
            {
                _todoContext.Remove(todo);
                _todoContext.SaveChanges();
            }
            catch (Exception)
            {

                throw;
            }
            return true;
        }

        public List<Models.Todo> FindAll()
        {
            var list = _todoContext.Todos.Select(t => new Models.Todo { Id = t.Id, Description = t.Description, CreatedAt = t.CreatedAt, CompletionDate = t.CompletionDate, Status = t.Status }).ToList();
            return list;
        }

        public Models.Todo? FindById(int id)
        {
            var todo = _todoContext.Todos.Where(t => t.Id == id).Select(t => new Models.Todo { Id = t.Id, Description = t.Description, CreatedAt = t.CreatedAt, CompletionDate = t.CompletionDate, Status = t.Status }).SingleOrDefault();
            return todo;
        }

        public Models.Todo? Update(Models.Todo todo)
        {
            var updateEntity = _todoContext.Todos.Where(t => t.Id == todo.Id).SingleOrDefault();
            if (updateEntity == null) return null;
            try
            {
                updateEntity.Description = todo.Description;
                updateEntity.CreatedAt = todo.CreatedAt;
                updateEntity.CompletionDate = todo.CompletionDate;
                updateEntity.Status = todo.Status;
                _todoContext.SaveChanges();
            }
            catch (Exception)
            {
                throw;
            }

            return todo;
        }
    }
}

File: todoApp/lib/repositories/todoRepository/todo_repository.dart
import 'package:todoapp/models/todo.dart';
import 'package:todoapp/repositories/todoRepository/todo_repository_interface.dart';
import 'package:todoapp/services/api/todo_api.dart';

class TodoRepository implements TodoRepositoryInterface {
  final _api = TodoApi();

  @override
  Future<bool> delete(int id) async {
    return await _api.delete(id);
  }

  @override
  Future<List<Todo>> findAll() async {
    return await _api.getAll();
  }

  @override
  Future<Todo?> findOne(int id) async {
    return await _api.getOne(id);
  }

  @override
  Future<Todo?> update(Todo todo) async {
    return await _api.put(todo);
  }

  @override
  Future<Todo?> create(Todo todo) async {
    return await _api.post(todo);
  }
}
